### PR TITLE
Fixed compilation issues with no outdir

### LINKED ISSUE #8280: error: cannot write multiple output files without an output directory
### What version of Bun is running?

1.0.23+83f2432da

### What platform is your computer?

Darwin 23.2.0 x86_64 i386

### What steps can reproduce the bug?

```ts
async function main() {
  const result = await Bun.build({
    entrypoints: ['./src/entry.ts'],
    naming: '[dir]/[name].[hash].[ext]',
    splitting: true,
  });

  console.log(result);
}
```
Gives:
```bash
{
  outputs: [],
  success: false,
  logs: [
    error: cannot write multiple output files without an output directory
  ],
}
```
Adding an `outDir` produces the following results:
Screenshot 2024-01-19 at 14 53 44

### What is the expected behavior?

I would expect that not having an `outDir` will produce exactly the same output, just not written to disk.

### What do you see instead?

The error above

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by BjornTheProgrammer**
### What does this PR do?

This PR fixes #8280, by changing
```zig
if (root_path.len == 0 and c.parse_graph.additional_output_files.items.len > 0 and !c.resolver.opts.compile) {
    try c.log.addError(null, Logger.Loc.Empty, "cannot write multiple output files without an output directory");
    return error.MultipleOutputFilesWithoutOutputDir;
}

```

```zig
const more_than_one_output = c.parse_graph.additional_output_files.items.len > 0 or c.parse_graph.generate_bytecode_cache or (has_css_chunk and has_js_chunk);

if (!c.resolver.opts.compile and more_than_one_output and (c.resolver.opts.outfile or c.resolver.opts.output_to_stdout)) {
    try c.log.addError(null, Logger.Loc.Empty, "cannot write multiple output files without an output directory");
    return error.MultipleOutputFilesWithoutOutputDir;
}
```

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I ran `./build/debug/bun-debug test bundle`

I wrote two automated bundler tests and extended the expectBundled to enable coverage of the issue this pr seeks to solve.

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)

### CODE REVIEW

File: test/bundler/bundler_edgecase.test.ts [hunk 1]
```diff
@@ -2031,6 +2031,41 @@ describe("bundler", () => {
...
+      `,
+      "/index.html": `
+        <head></head>
+      `,
+    },
+    generateOutput: false,
+    backend: "api",
+    onAfterApiBundle: build => {
+      expect(build.success).toEqual(true);
+      expect(build.outputs).toBeArrayOfSize(2);
```
Comment (paperclover) [inline test/bundler/bundler_edgecase.test.ts:2048]: can you verify that the build.outputs items are correct? mainly if you read the outputted `index.html`, you get the text. maybe also check if it's content type is right.

File: src/options.zig [hunk 2]
```diff
@@ -1529,6 +1529,9 @@ pub const BundleOptions = struct {
     /// So we have a list of packages which we know are safe to do this with.
     unwrap_commonjs_packages: []const string = &default_unwrap_commonjs_packages,
 
+    outfile: bool = false,
```
Comment (paperclover) [inline src/options.zig]: outfile being a boolean is kind of silly, can we do without this?

File: src/options.zig [hunk 3]
```diff
@@ -1529,6 +1529,9 @@ pub const BundleOptions = struct {
...
+    output_to_stdout: bool = false,
```
Comment (paperclover) [inline src/options.zig]: then this could be changed to `supports_multiple_outputs`

**Review by paperclover — CHANGES_REQUESTED** [review]
this seems mostly reasonable. have a request to change how the information is passed from build_command to bundle_v2 and one nit on the test

File: test/bundler/bundler_edgecase.test.ts (hunk 1 above)
Comment (BjornTheProgrammer) [reply test/bundler/bundler_edgecase.test.ts:2048]: Makes sense, have done so in newest commit!

File: src/options.zig (hunk 2 above)
Comment (BjornTheProgrammer) [reply src/options.zig]: You're totally right, folded this into the variable supports_multiple_outputs

**Review by paperclover — APPROVED** [review]

**Comment by BjornTheProgrammer** [comment]
I've rebased to the latest, and tested to ensure everything still works.

**Comment by BjornTheProgrammer** [comment]
I'm really confused by these build tests. Are the tests just failing sporadically, also why does it seem like some just don't run? Is it an issue with my code?

**Pull request merged**